src/Core/Checkout/Document/Renderer/InvoiceRenderer.php

```diff
@@ -140,17 +147,36 @@ public function render(array $operations, Context $context, DocumentRendererConf
                         $context,
                         $order->getSalesChannelId(),
                         $order->getLanguageId(),
-                        $locale->getCode()
+                        $locale->getCode(),
                     );
 
                     $doc = new RenderedDocument(
-                        $html,
+                        $html, // @deprecated tag:v6.7.0 - html argument will be removed
                         $number,
                         $config->buildName(),
                         $operation->getFileType(),
                         $config->jsonSerialize(),
                     );
 
+                    // set the template renderer to be able to render template
+                    $doc->setTemplateOptions([
+                        $template,
+                        [
+                            'order' => $order,
+                            'config' => $config,
+                            'rootDir' => $this->rootDir,
+                            'context' => $context,
+                        ],
+                        $context,
+                        $order->getSalesChannelId(),
+                        $order->getLanguageId(),
+                        $locale->getCode(),
+                    ]);
```

Comment:
why do we build such an ugly array, where to get the localeCode, I have to do `$doc->getTemplateOptions()[5]`?
Why not have typed properties for all parameters? Or at least an object?

Reply:
Another problem with this is, what if one of these properties aren't included because it's not needed, then the index changes and now you are accessing the wrong index.

Reply:
In other words, I would pass `$template`, `$order`, `$config` and `$locale` into the `RenderedDocument` class and assemble the signature in the TypeRenderers, where it is actually needed to call the `DocumentTemplateRenderer`.

`$this->rootDir` can be injected in the TypeRenderers, no need to pass it into the `RenderedDocument`.

Benefit of all this is that if a renderer needs access to the Context or the OrderEntity and does not use the Twig Engine, it can do so easily.